crates/fuzzing/src/generators/value.rs

```diff
@@ -0,0 +1,144 @@
+
+use arbitrary::{Arbitrary, Unstructured};
+use std::hash::Hash;
+
+/// A value passed to and from evaluation. Note that reference types are not
+/// (yet) supported.
+///
+/// TODO implement correct partial equality comparisons for FP.
+#[derive(Clone, Debug, PartialEq)]
+#[allow(missing_docs)]
+pub enum DiffValue {
+    I32(i32),
+    I64(i64),
+    F32(u32),
+    F64(u64),
+    V128(u128),
+}
+
+impl DiffValue {
+    fn ty(&self) -> DiffValueType {
+        match self {
+            DiffValue::I32(_) => DiffValueType::I32,
+            DiffValue::I64(_) => DiffValueType::I64,
+            DiffValue::F32(_) => DiffValueType::F32,
+            DiffValue::F64(_) => DiffValueType::F64,
+            DiffValue::V128(_) => DiffValueType::V128,
+        }
+    }
+
+    /// Generate a [`DiffValue`] of the given `ty` type.
+    ///
+    /// This function will bias the returned value 50% of the time towards one
+    /// of a set of known values (e.g., NaN, -1, 0, infinity, etc.).
+    pub fn arbitrary_of_type(
+        u: &mut Unstructured<'_>,
+        ty: DiffValueType,
+    ) -> arbitrary::Result<Self> {
+        use DiffValueType::*;
+        let val = match ty {
+            I32 => DiffValue::I32(biased_arbitrary_value(u, KNOWN_I32_VALUES)?),
+            I64 => DiffValue::I64(biased_arbitrary_value(u, KNOWN_I64_VALUES)?),
+            F32 => {
+                // TODO once `to_bits` is stable as a `const` function, move
+                // this to a `const` definition.
+                let known_f32_values = &[
+                    f32::NAN.to_bits(),
+                    f32::INFINITY.to_bits(),
+                    f32::NEG_INFINITY.to_bits(),
+                    f32::MIN.to_bits(),
+                    (-1.0f32).to_bits(),
+                    (0.0f32).to_bits(),
+                    (1.0f32).to_bits(),
+                    f32::MAX.to_bits(),
+                ];
+                DiffValue::F32(biased_arbitrary_value(u, known_f32_values)?)
+            }
+            F64 => {
+                // TODO once `to_bits` is stable as a `const` function, move
+                // this to a `const` definition.
+                let known_f64_values = &[
+                    f64::NAN.to_bits(),
+                    f64::INFINITY.to_bits(),
+                    f64::NEG_INFINITY.to_bits(),
+                    f64::MIN.to_bits(),
+                    (-1.0f64).to_bits(),
+                    (0.0f64).to_bits(),
+                    (1.0f64).to_bits(),
+                    f64::MAX.to_bits(),
+                ];
+                DiffValue::F64(biased_arbitrary_value(u, known_f64_values)?)
+            }
+            V128 => DiffValue::V128(biased_arbitrary_value(u, KNOWN_U128_VALUES)?),
+        };
+        arbitrary::Result::Ok(val)
+    }
+}
+
+const KNOWN_I32_VALUES: &[i32] = &[i32::MIN, -1, 0, 1, i32::MAX];
+const KNOWN_I64_VALUES: &[i64] = &[i64::MIN, -1, 0, 1, i64::MAX];
+const KNOWN_U128_VALUES: &[u128] = &[u128::MIN, 1, u128::MAX];
```

Comment:
One thing that might be good to add here is "known values" for each lane width like (1u64, 1u64), (-1i64, -1i64), (1u32, 2u32, 3u32, 4u32), etc.